File: tests/benchmarks/aot_c90/aggregates/class_direct_method_loop.c
#include <stdint.h>
#include <stdio.h>

typedef struct {
    int64_t value;
    int64_t step;
} Counter;

static void counter_init(Counter *c, int64_t init, int64_t step) {
    c->value = init;
    c->step = step;
}

static int64_t counter_bump(Counter *c, int64_t n) {
    int64_t i = 0;
    int64_t sum = 0;
    while (i < n) {
        c->value = c->value + c->step;
        sum = sum + c->value;
        i = i + 1;
    }
    return sum + c->value;
}

static int64_t run(int64_t n) {
    Counter c;
    counter_init(&c, 1, 3);
    return counter_bump(&c, n);
}

int main(void) {
    printf("%lld\n", (long long) run(2000000));
    return 0;
}

File: tests/benchmarks/aot_c90/aggregates/class_direct_method_loop.rs
// Rust reference for class_direct_method_loop.xr (statement-level isomorphic, safe std-only).

struct Counter {
    value: i64,
    step: i64,
}

impl Counter {
    fn new(init: i64, step: i64) -> Self {
        Counter {
            value: init,
            step,
        }
    }

    fn bump(&mut self, n: i64) -> i64 {
        let mut i: i64 = 0;
        let mut sum: i64 = 0;
        while i < n {
            self.value += self.step;
            sum += self.value;
            i += 1;
        }
        sum + self.value
    }
}

fn run(n: i64) -> i64 {
    let mut c = Counter::new(1, 3);
    c.bump(n)
}

fn main() {
    println!("{}", run(2000000));
}
